contrib/scalar/scalar.c

```diff
@@ -19,9 +19,27 @@ static int is_unattended(void) {
 	return git_env_bool("Scalar_UNATTENDED", 0);
 }
 
+static void strbuf_parentdir(struct strbuf *buf)
+{
+	int len = buf->len;
+	while (len > 0 && !is_dir_sep(buf->buf[--len]))
+		; /* keep looking for parent directory */
+	strbuf_setlen(buf, len);
+}
+
+/* Given an absolute working directory, find the enlistment root */
+static void strbuf_enlistmentroot(struct strbuf *buf)
+{
+	int len = buf->len;
+	if (len >= 4 && is_dir_sep(buf->buf[len-4]) &&
+			!strncmp("src", buf->buf+len-3, 3))
+		strbuf_parentdir(buf);
+}
```

Comment:
I am not quite sure that I understand what `strbuf_enlistment_root()` is supposed to do. For example, if you are in `blub/out` and there exists a `blub/src/.git`, shouldn't we return `blub`?

If so, I would like to suggest three things:

- Declare `size_t len = buf->len;`
- Declare the return type `int` and return `-1` if no enlistment could be found
- Use a `for (;;)` loop which starts by appending `/src/.git` and testing whether it is a Git directory via `is_git_directory()`, returning 0 if it is
- In the loop, search backwards for the next directory separator (using `is_dir_sep(len)`, taking care not to look further than `offset_1st_component(buf->buf)`.
- Use `strbuf_setlen()` to truncate the path in the loop when finding a directory separator (or hitting the offset of the 1st component).

Reply:
`strbuf_enlistment_root()` was (mostly) refactored into `find_enlistment(...)`, but hopefully this can clarify what I was going for:

- Originally (before this change), `setup_enlistment_directory(...)` would find the git working directory (_not_ enlistment root) relevant to the command and `chdir` to that working directory. The implicit assumption (made throughout the rest of the file) was that the enlistment root was _always_ the parent of the working directory, and the working directory was named `src`
- Here, `setup_enlistment_directory(...)` (and its internal logic, refactored into `find_enlistment(...)`) still performs the same function (find the working directory, `chdir` to it), but it also takes an optional `strbuf *enlistment_root` argument. If provided (as an empty buffer), it is set to the enlistment root. The enlistment root is the parent of the working directory if the working directory is `src`, otherwise it is the working directory itself. For example:
  - Working directory: `/usr/src/test-repo/src`; enlistment root: `/usr/src/test-repo`
  - Working directory: `/usr/src/another-repo`; enlistment root: `/usr/src/another-repo`

Because the input path for `setup_enlistment_directory(...)` could be any path under the enlistment root, setting `enlistment_root` in-place seemed like a relatively non-invasive way of getting additional (but not always required) information about the enlistment. 

In addition to that, this implementation handles some edge cases that existed with finding the working directory in the first place - for example, calling `scalar register .` from inside a working directory that contained a `src` folder would falsely register `workdir/src` as the working directory.

Reply:
Right, we should really make sure to find a worktree there, not just an `src` directory.

I still think we need to be careful to find the enlistment root directory when the current directory is, say, `<root>/out/` (and the repository is in `<root>/src/.git`).

Reply:
`find_enlistment(...)` ensures `src/.git` exists and is a git directory, so that should address your first concern. For the second, I've updated the implementation to handle that case (as well as added a test covering that scenario).